forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/listeners/FormPDFListener.java

```diff
@@ -0,0 +1,319 @@
+import org.camunda.bpm.extension.mail.send.SendMailInvocation;
+import org.camunda.bpm.extension.mail.send.SendMailRequest;
+import org.camunda.bpm.extension.mail.service.MailService;
+import org.camunda.bpm.extension.mail.service.MailServiceFactory;
+import org.camunda.connect.impl.AbstractConnector;
+import org.camunda.connect.spi.ConnectorRequestInterceptor;
+import org.camunda.connect.spi.ConnectorResponse;
+import org.json.JSONObject;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+
+import javax.activation.DataHandler;
+import javax.activation.DataSource;
+import javax.activation.FileDataSource;
+import javax.inject.Named;
+import javax.mail.*;
+import javax.mail.internet.InternetAddress;
+import javax.mail.internet.MimeBodyPart;
+import javax.mail.internet.MimeMessage;
+import javax.mail.internet.MimeMultipart;
+import javax.ws.rs.core.Application;
+import java.io.*;
+import java.nio.file.Paths;
+import java.util.*;
+
+@Named("FormPDFListener")
+public class FormPDFListener extends BaseListener implements TaskListener, IMessageEvent {
+
+    protected MailConfiguration configuration;
+
+    private String formUrl;
+    private String managerEmail;
+    private String submitterEmail;
+
+    @Autowired
+    private FormSubmissionService formSubmissionService;
+
+    @Autowired
+    private HTTPServiceInvoker httpServiceInvoker;
+
+    @Override
+    public void notify(DelegateTask delegateTask) {
+        try {
+
+            // Get Form Values
+            DelegateExecution execution = delegateTask.getExecution();
+            String formUrlString = String.valueOf(execution.getVariables().get("formUrl"));
+            JSONObject formValues = formSubmissionService.retrieveFormJson(formUrlString).getJSONObject("data");
+
+            // Get the URL for the form fields
+            formUrl = formUrlString.split("/submission/")[0];
+            JSONObject formFields = formSubmissionService.retrieveFormJson(formUrl);
+
+            // Extract the emails for the manager and submitter
+            Map<String, Object> valuesDataMap = formSubmissionService.retrieveFormValues(formUrlString);
+            managerEmail = String.valueOf(valuesDataMap.get("managerEmail"));
+            submitterEmail = String.valueOf(valuesDataMap.get("email"));
+
+            // Create the data to be passed to the template
+            Map<String, Object> data = new HashMap<>();
+            data.put("formFields", formFields );
+            data.put("formValues", formValues);
+
+            // Generate the template from '/templates/form.html'
+            Template template = getTemplate();
+
+            // Save the template as output.html
+            Writer fileWriter = new FileWriter(new File("output.html"));
+            template.process(data, fileWriter);
```

Comment:
Ideally we would write this to a temporary file so avoid any chance of `output.html` being reused by someone else for example two requests coming in at the same time.

```java
File f = null;
try {
        f = File.createTempFile();
        template.process(data, new FileWriter(f));
        RenderPage(f);
} finally {
     if(f != null) {
       f. deleteIfExists();
     }
}      

```

Reply:
pro tip :)